### ISSUE TITLE
torch.logdet produces incorrect results for singular matrices on CUDA vs CPU

### ISSUE THREAD

**Issue by ChaitanyaRS06** [pytorch/pytorch#154312]
### 🐛 Describe the bug

torch.logdet() returns mathematically incorrect results when computing the log determinant of singular matrices on CUDA devices, while CPU computation is correct.

For a singular matrix (determinant = 0), the log determinant should be -∞. CPU correctly returns -inf, but CUDA returns a finite value.

```python
import torch

# Singular matrix (Row 3 = 3 * Row 1)
matrix = torch.tensor([[1.0, 2.0, 3.0], 
                       [2.0, 5.0, 6.0], 
                       [3.0, 6.0, 9.0]], dtype=torch.float32)

cpu_result = matrix.logdet()
gpu_result = matrix.to('cuda').logdet()

print(f"CPU: {cpu_result}")      # -inf (correct)
print(f"GPU: {gpu_result}")      # -15.1315 (incorrect)
print(f"Determinant: {matrix.det()}")  # Should be ~0
```

Output:
```
CPU: -inf
GPU: -15.131454467773438
Determinant: -0.0
```

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Rocky Linux release 8.9 (Green Obsidian) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-20)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.28

Python version: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-513.18.1.el8_9.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB

Nvidia driver version: 550.54.14
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              128
On-line CPU(s) list: 0-127
Thread(s) per core:  1
Core(s) per socket:  64
Socket(s):           2
NUMA node(s):        8
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7742 64-Core Processor
Stepping:            0
CPU MHz:             2250.000
CPU max MHz:         2250.0000
CPU min MHz:         1500.0000
BogoMIPS:            4491.48
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-15
NUMA node1 CPU(s):   16-31
NUMA node2 CPU(s):   32-47
NUMA node3 CPU(s):   48-63
NUMA node4 CPU(s):   64-79
NUMA node5 CPU(s):   80-95
NUMA node6 CPU(s):   96-111
NUMA node7 CPU(s):   112-127
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] pytorch-fid==0.3.0
[pip3] torch==2.6.0
[pip3] torchaudio==2.6.0
[pip3] torchmetrics==1.6.0
[pip3] torchvision==0.15.2+cu118
[pip3] triton==3.2.0
[conda] numpy  

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @ptrblck @eqy @jerryzh168 @jianyuh @nikitaved @mruberry @walterddr @xwang233 @Lezcano

**Comment by jbschlosser** [comment]
The next part of the statement from the numerical accuracy docs says:
```
"... when the inputs have non-finite values like inf or NaN"
```

In this case, inputs are all non-inf and non-NaN. The docs for `logdet()` explicitly say it will return -inf for inputs with determinant == 0 so to me this reads as a GPU-side correctness issue.

**Comment by redwrasse** [comment]
My interpretation was more along the lines that `logdet` as a matrix function is ill-conditioned (infinite condition number) about singular matrix inputs. But if the specific API method docs are a source of truth for the expected behavior, this is great.

**Comment by soumith** [comment]
Put up a fix here https://github.com/pytorch/pytorch/pull/157910

**Comment by jiren-the-gray** [comment]
I can still reproduce this bug in `torch==2.9.0`. Here is a [gist](https://colab.research.google.com/gist/jiren-the-gray/25d259efe201c62cdbb14f3e8be991bd/logdet_bug.ipynb).

Can this issue be re-opened or the PR merged properly?

@malfet @jbschlosser @soumith
